Book.java

```diff
@@ -0,0 +1,315 @@
+     * relevant to that type of contract.
+     * @param contract a Liability contract to add
+     */
+    public void addLiability(Contract contract) {
+        Account liabilityAccount = liabilityAccountsMap.get(contract.getClass());
+
+
+        if (liabilityAccount==null) {
+            // If there doesn't exist an Account to hold this type of contract, we create it
+            liabilityAccount = new Account(contract.getClass().getName(), AccountType.LIABILITY);
+            addAccount(liabilityAccount, contract.getClass());
+        }
+
+        // (dr equity, cr liability)
+        equityAccount.debit(contract.getValue());
+        liabilityAccount.credit(contract.getValue());
+
+        liabilityAccount.addContract(contract);
+    }
+
+    public void addCash(double amount) {
+
+        // (dr cash, cr equity)
+        cashAccount.debit(amount);
+        equityAccount.credit(amount);
+    }
+
+
+    /**
+     * Operation to cancel a Loan to someone (i.e. cash in a Loan in the Assets side).
+     *
+     * I'm using this for simplicity but note that this is equivalent to selling an asset.
+     * @param amount the amount of loan that is cancelled
+     */
+    public void pullFunding(double amount) {
+        Account loanAccount = assetAccountsMap.get(Loan.class);
+
+        // (dr cash, cr asset )
+        cashAccount.debit(amount);
+        loanAccount.credit(amount);
+
+    }
+
+    /**
+     * Operation to pay back a liability loan; debit liability and credit cash
+     * @param amount amount to pay back
+     */
+    public void payLiability(double amount, Class<? extends Contract> liabilityType) {
+        Account liabilityAccount = liabilityAccountsMap.get(liabilityType);
+
+        //Todo: What do we do if we can't pay??!! At the moment I'm calling my owner to raise liquidity
+        if (cashAccount.getBalance() < amount) {
+            System.out.println();
+            System.out.println("***");
+            System.out.println(owner.getName()+" must raise liquidity immediately.");
+            owner.raiseLiquidity(amount * (1 - cashAccount.getBalance()/getAssetValue()));
+            System.out.println("***");
+            System.out.println();
+        }
+
+        // (dr liability, cr cash )
+        liabilityAccount.debit(amount);
+        cashAccount.credit(amount);
+
+    }
+
+    /**
+     * If I've sold an asset, debit cash and credit asset
+     * @param amount the *value* of the asset
+     */
+    public void sellAsset(double amount, Class<? extends Contract> assetType) {
+        Account assetAccount = assetAccountsMap.get(assetType);
+
+        // (dr cash, cr asset)
+        cashAccount.debit(amount);
+        assetAccount.credit(amount);
+    }
+
+    /**
+     * Behavioral stuff; not sure if it should be here
+     * @param me the owner of the Book
+     * @return an ArrayList of Actions that are available to me at this moment
+     */
+    public ArrayList<Action> getAvailableActions(Agent me) {
```

Comment:
Let's discuss actions in the code review. I really like the idea. Perhaps we should have some kind of interface which defines a list of actions, since we'd probably use them in a variety of contexts.